File: python/0977.py
# Link: https://leetcode.com/problems/squares-of-a-sorted-array/description/
# Runtime: 145 ms
# Memory: 18.89 MB


class Solution:
    def sortedSquares(self, nums: List[int]) -> List[int]:
        sorted_squares_array = nums[:]
        left, right = 0, len(nums) - 1
        for i in range(len(nums) - 1, -1, -1):
            if abs(nums[left]) < abs(nums[right]):
                sorted_squares_array[i] = nums[right] * nums[right]
                right -= 1
            else:
                sorted_squares_array[i] = nums[left] * nums[left]
                left += 1

        return sorted_squares_array

File: c++/0977.cpp
// Link: https://leetcode.com/problems/squares-of-a-sorted-array/description/
// Runtime: 11 ms
// Memory: 28.28 MB

class Solution {
public:
  std::vector<int> sortedSquares(const std::vector<int> &nums) {
    std::vector<int> sortedSquaresArray(nums.size());
    int left = 0, right = static_cast<int>(nums.size() - 1);
    for (int i = static_cast<int>(nums.size() - 1); i > -1; i--) {
      if (abs(nums[left]) < abs(nums[right])) {
        sortedSquaresArray[i] = nums[right] * nums[right];
        right--;
      } else {
        sortedSquaresArray[i] = nums[left] * nums[left];
        left++;
      }
    }

    return sortedSquaresArray;
  }
};
